File: python/queen-attack/queen_attack.py
def position_valid(row, column):
    # if the row parameter is negative
    if row < 0:
        raise ValueError("row not positive")
    # if the row parameter is not on the defined board
    if row > 7:
        raise ValueError("row not on board")
    # if the column parameter is negative
    if column < 0:
       raise ValueError("column not positive")
    # if the column parameter is not on the defined board
    if column > 7:
        raise ValueError("column not on board")
    
    return True

class Queen:
    # a Queen object has an attribute position, which is a tuple of two integers
    position = (0, 0)

    # initialize the queen's position using the row and column parameters
    def __init__(self, row, column):
        if position_valid(row, column):
            self.position = (row, column)

    def can_attack(self, another_queen):
        # if the queens are in the same square
        if self.position == another_queen.position:
            raise ValueError("Invalid queen position: both queens in the same square")
        
        # if the queens are on the same row
        if self.position[0] == another_queen.position[0]:
            return True

        # if the queens are on the same column
        if self.position[1] == another_queen.position[1]:
            return True

        # if the queens are on the same diagonal
        slope = (self.position[0] - another_queen.position[0]) / (self.position[1] - another_queen.position[1])
        return True if slope == 1 or slope == -1 else False

File: c/queen-attack/queen_attack.c
#include "queen_attack.h"

attack_status_t can_attack(position_t queen_1, position_t queen_2)
{
    // Check if a queen is in an invalid position
    if (!is_valid_position(queen_1) || !is_valid_position(queen_2) || (queen_1.row == queen_2.row && queen_1.column == queen_2.column))
    {
        return INVALID_POSITION;
    }

    int slope = 0;
    int row_diff = queen_2.row - queen_1.row;
    int col_diff = queen_2.column - queen_1.column;
    if (col_diff != 0)
    {
        slope = row_diff / col_diff;
    }
    else
    {
        // Since col_diff is 0, the queens can attack.
        return CAN_ATTACK;
    }

    // If the queens are on the same row, column, or diagonal, then they can attack.
    if (row_diff == 0 || col_diff == 0 || slope == 1 || slope == -1)
    {
        return CAN_ATTACK;
    }
    else
    {
        return CAN_NOT_ATTACK;
    }
}

bool is_valid_position(position_t queen)
{
    // A queen's position is invalid if it is not in the range 0 to 7.
    // The data type uint8_t is unsigned and thus always starts at 0, we no longer have to check for that condition.
    return queen.row <= 7 && queen.column <= 7;
}